### PR TITLE
added autofix on save

### PR DESCRIPTION
**Pull request by wimva**
Inspired on https://github.com/AtomLinter/linter-stylelint/pull/347

This was not what I expected from the fix on save feature, so I rewrote some parts so it would work as expected.

### CODE REVIEW

**Comment by nicksmetBagaar** [comment]
+1

**Comment by ctf0** [comment]
@Arcanemagus any news on this ?

**Comment by lydell** [comment]
I just installed this (`apm install wimva/linter-stylelint`), and so far it seems to work. But every time I save the cursor ends up at the bottom of the file. Would be awesome if the cursor stayed roughly where it were, like in linter-eslint!

**Comment by wimva** [comment]
Yes your right, I’ll try to fix this.

**Comment by wimva** [comment]
@lydell This should be fixed now.

**Comment by lydell** [comment]
Thanks @wimva, that seems to work!

Another little issue, this time regarding undo:

1. Make some valid edit, such as changing `red` to `blue`.
2. Make a change that violates an autofixable stylelint rule, such as chaning to the wrong type of quotes.
3. Save. The file is autofixed as expected.
4. Press ctrl+z to undo. Nothing appears to happen.
5. Press ctrl+z to undo again. Now the autofix is undone.
6. Press ctrl+z to undo one last time. Now the first edit (such as `red` to `blue`) is undone.

In other words, undo works as expected, _except_ step 4. After saving, ctrl+z should undo the autofix directly – I should not have to press ctrl+z twice. This is how linter-eslint works.

**Comment by lydell** [comment]
I've been using this for a couple of days at work now, and it works really great! Thanks!

The undo issue is the only thing I've run into, and IMO it is not at all critical. @wimva I saw you reacted with the "confused" emoji – is there anything you'd like me to clarify? :)

If the maintainers find some technical edge case or so I'd say _ship it_ and fix it later! :shipit: :tada:

**Comment by wimva** [comment]
@lydell I reacted with the confused emoji because I don't really know how to fix that one at the moment. I'll have to dive deeper into it. If you have any idea, you're welcome to help :)

**Comment by remhume** [comment]
Can this be merged then? @Arcanemagus

As far as i can see the last comments seem to agree that this is shippable(?), and i'm a bit confused nothing has happened here as this is a feature i'm quite excited about :)
(I do not at all want to complain though, lives are full and stuff, thanks for your work @wimva and the other contributors <3)

I'm totally pro shipping and fixing that specific mentioned bug later when there is more time and/or help, as this seems to be a minor detail.

**Comment by wimva** [comment]
@remhume You can always run `apm install wimva/linter-stylelint` to use my version. I'll try to keep this updated and when I find some more time I will look in the undo bug mentioned above. I really enjoy this auto fixing and it saves a lot of time :)

**Comment by Arcanemagus** [comment]
Would you mind rebasing this on `master` and updating the commit messages to follow [Conventional Commits](https://conventionalcommits.org/)? Will be reviewing this in more detail later tonight, and sorry for the delay!

**Comment by wimva** [comment]
I was on holiday last week, but I will check this asap!

**Comment by wimva** [comment]
@Arcanemagus updated, should be ok now :)

File: lib/index.js [hunk 1]
```diff
@@ -77,13 +87,46 @@ export default {
...
+  },
+
   provideLinter() {
     return {
       name: 'stylelint',
       grammarScopes: this.baseScopes,
       scope: 'file',
       lintsOnChange: true,
-      lint: async (editor) => {
+      lint: async (editor, autofix) => {
```
Comment (Arcanemagus) [inline lib/index.js]: I would name this `forceFix`, and default it to false here.

File: lib/index.js [hunk 2]
```diff
@@ -99,7 +142,8 @@ export default {
 
         const options = {
           code: text,
-          codeFilename: filePath
+          codeFilename: filePath,
+          fix: !!autofix
```
Comment (Arcanemagus) [inline lib/index.js]: This should be `this.autofix || forceFix`.

File: lib/index.js [hunk 3]
```diff
@@ -77,13 +87,46 @@ export default {
     this.subscriptions.dispose();
   },
 
+  fixJob(editor) {
+    let correctScope = false;
+
+    if (editor && editor.buffer && editor.buffer.file && editor.buffer.file.path) {
```
Comment (Arcanemagus) [inline lib/index.js]: Use [Workspace::isTextEditor](https://atom.io/docs/api/v1.26.1/Workspace#instance-isTextEditor) and [TextEditor::getPath](https://atom.io/docs/api/v1.26.1/TextEditor#instance-getPath).

File: lib/index.js [hunk 4]
```diff
@@ -77,13 +87,46 @@ export default {
...
+      const { path } = editor.buffer.file;
+
+      this.baseScopes.forEach((scope) => {
```
Comment (Arcanemagus) [inline lib/index.js]: Use logic [like this](https://github.com/AtomLinter/linter-eslint/blob/62b44afa4f5d876b6232b3dac04aba1081d8ff2a/src/validate/editor.js#L18-L21) to check if the `editor` has a cursor that is within a location that has a valid scope.

File: lib/index.js [hunk 5]
```diff
@@ -77,13 +87,46 @@ export default {
...
+    }
+
+    if (!correctScope) {
+      return;
+    }
+
+    const oldFileContent = editor.getText();
+    const cursorPosition = editor.getCursorBufferPosition();
+
+    this.provideLinter().lint(editor, true).then(() => {
```
Comment (Arcanemagus) [inline lib/index.js]: This will actually initialize a _new_ provider, I'm not sure how problematic that would be, but maybe `provideLinter` should cached the object in `this` (so package activation/deactivation still gets a new instance)?

File: lib/index.js [hunk 6]
```diff
@@ -77,13 +87,46 @@ export default {
...
+      const newFileContent = editor.getText();
+
+      if (newFileContent !== oldFileContent) {
+        editor.save();
```
Comment (Arcanemagus) [inline lib/index.js]: Instead of saving the file twice, use [TextBuffer::onWillSave](https://atom.io/docs/api/v1.26.1/TextBuffer#instance-onWillSave) to make Atom wait for the save job to complete before saving the file to disk. Note that you will need to make the observation of the `TextEditor`s above this more complex since that method only (publicly) exists on the `TextBuffer`.

File: lib/index.js [hunk 7]
```diff
@@ -77,13 +87,46 @@ export default {
...
+        }
+      });
+    }
+
+    if (!correctScope) {
+      return;
+    }
+
+    const oldFileContent = editor.getText();
+    const cursorPosition = editor.getCursorBufferPosition();
```
Comment (Arcanemagus) [inline lib/index.js]: The way you currently have this written it will automatically consolidate all `Cursor`s down to a single one at the position of the last added one when fixing the document (regardless of whether anything changed!).

I really like the idea of attempting to restore the cursor position, so here are two ways I see of fixing this to support multiple cursors:

* The quick way of just leaving it as is, and forcing users to accept it
* The "right" way of tracking all cursors and setting the position of each one


If we want to track all the `Cursor`s I _think_ it would look something like:
1. Grab all the `Cursor`s with [TextEditor::getCursors](https://atom.io/docs/api/v1.26.1/TextEditor#instance-getCursors)
2. Store them in a `Map`, with the `Cursor` as the key, and the current position as the value
3. Run the fix job
4. Attempt to set each `Cursor`'s position to the cached position

A potential issue I see with this is what will Atom do to the existing `Cursor`s when we call `editor.setText`? If it consolidates them together then, _destorying_ the old ones, we may need to create new `Cursor`'s at the old locations in the older they were added which would complicate things further.

File: lib/index.js (hunk 6 above)
Comment (lydell) [reply lib/index.js]: Saving the file twice? Perhaps that fixes the undo issue I posted about…

Comment (Arcanemagus) [reply lib/index.js]: Yep, this would be the cause of that problem. As the code works currently it is letting Atom save the file, then running the fix job, updating the text, and then saving it again.

**Comment by Arcanemagus** [comment]
@wimva Did any of the above not make sense? Need further pointers on how to implement them?

**Comment by wimva** [comment]
@Arcanemagus I did not have time yet to go through it all, but I suppose all is clear :). As I'm very busy at the moment it will take some time to get the change requests done, but I will have a look as soon as possible!

**Comment by Braintelligence** [comment]
Will this be in review soon? 😸

**Comment by IanVS** [comment]
@wimva, I'd also love to see this land.  Is there anything I can do to help out?  If you don't have time to finish up, would you be opposed to me cherry-picking your commits into my own branch and running with them?

**Comment by wimva** [comment]
@IanVS I currently am too busy to further work on this. If you like and have time, please cherry pick and fix open issues.

**Comment by alexwendte** [comment]
oooo I hope this comes out soon! This looks like it would be so helpful 😁

**Comment by IanVS** [comment]
@alexwendte, if you'd like to checkout my branch in https://github.com/AtomLinter/linter-stylelint/pull/418 and do an `apm uninstall linter-stylelint && apm link` from within it, you should be able to have this functionality (after enabling the option).  And you can give feedback on how well it is working for you, or if there are any bugs.
